features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+//     type: "recovery_code",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//   },
+//   {
+//     id: "5",
+//     type: "recovery_code",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//   },
+//   {
+//     id: "6",
+//     type: "bearer_token",
+//     createdAt: new Date("2019-07-19T00:00:00.000Z"),
+//     expireAt: new Date("2019-08-19T00:00:00.000Z"),
+//   }
+// ]
+
+const numberOfRecoveryCodes = authenticators.filter(a => a.type === "recovery_code").length;
+const hasTOTP = authenticators.filter(a => a.type === "totp").length > 0;
+const hasOOB = authenticators.filter(a => a.type === "oob").length > 0;
+```
+
+## Delete an authenticator
+
+```typescript
+const authenticators = await skygear.auth.mfa.listAuthenticators();
+// Present an UI to show the list of authenticators and
+// let the user to choose which one to delete.
+
+// Suppose the user wants to delete the first one.
+const { id } = authenticators[0];
+await skygear.auth.mfa.deleteAuthenticator(id);
+
+// The user can also delete a bearer token if they believe
+// the device storing it has been compromised.
+const bearerTokens = authenticators.filter(a => a.type === "bearer_token");
+const { id } = bearerTokens[0];
+await skygear.auth.mfa.deleteAuthenticator(id);
+```
+
+## Regenerate recovery codes
+
+```typescript
+const { recoveryCodes } = await skygear.auth.mfa.regenerateRecoveryCode();
+// Present an UI to show the list of recovery codes to the user.
+```
+
+# Access Token
+
+The access token includes a new claim [amr](https://openid.net/specs/openid-connect-core-1_0.html).
+
+## Mapping between existing authentication method and factor name
+
+|Method|Factor Name|
+|------|-----------|
+|password|password|
+|oauth|oauth|
+|custom_token|custom_token|
+
+When the user authenticates with one factor, the value is singleton array of the name of that factor.
+
+When the user authenticates with MFA factor, the authenticator type is added.
+
+## Example
+
+### Authenticate with password
+
+```JSON
+{
+  "amr": ["password"]
+}
+```
+
+### Authenticate with password and then TOTP
+
+```JSON
+{
+  "amr": ["password", "totp"]
+}
+```
+
+### Authenticate with password and then OOB
+
+```JSON
+{
+  "amr": ["password", "oob"]
+}
+```
+
+### Authenticate with password and then bearer token
+
+```JSON
+{
+  "amr": ["password", "bearer_token"]
+}
+```
+
+### Authenticate with password and then recovery code
+
+```JSON
+{
+  "amr": ["password", "recovery_code"]
+}
+```
+
+### Authenticate with OAuth and then TOTP
+
+```JSON
+{
+  "amr": ["oauth", "totp"]
+}
+```
+
+## Injection of amr into HTTP headers
+
+The gateway injects the value of `amr` into HTTP header `x-skygear-auth-amr` as a comma separated value.
+
+### Example
+
+Given
+
+```JSON
+{
+  "amr": ["password", "oob"]
+}
+```
+
+The injected header is
+
+```
+x-skygear-auth-amr: password, oob
+```
+
+The usefulness of this injection is to allow microservice to determine whether MFA is required.
```

Comment:
I think you mean determine whether the user is logged in with MFA?

Reply:
Oops. Typo.